### PR TITLE
CDI ref guide - document eager instantiation of beans

### PR DESCRIPTION
**Pull request by mkouba**

### CODE REVIEW

File: docs/src/main/asciidoc/cdi-reference.adoc [hunk 1]
```diff
@@ -168,6 +168,89 @@ public class CounterBean {
 
 == Non-standard Features
 
+=== Eager Instantiation of Beans
+
+[[lazy_by_default]]
+==== Lazy By Default
+
+By default, CDI beans are created lazily, when needed.
+What exactly means "needed" depends on the scope of a bean.
```
Comment (geoand) [inline docs/src/main/asciidoc/cdi-reference.adoc]: `means "needed"` -> `"needed" means`

File: docs/src/main/asciidoc/cdi-reference.adoc [hunk 2]
```diff
@@ -168,6 +168,89 @@ public class CounterBean {
...
+
+* A *normal scoped bean* (`@ApplicationScoped`, `@RequestScoped`, etc.) is needed when a method is invoked upon an injected instance (contextual reference per the specifitation).
```
Comment (geoand) [inline docs/src/main/asciidoc/cdi-reference.adoc]: `specifitation` -> `specification`

File: docs/src/main/asciidoc/cdi-reference.adoc [hunk 3]
```diff
@@ -168,6 +168,89 @@ public class CounterBean {
...
+
+[[lazy_by_default]]
+==== Lazy By Default
+
+By default, CDI beans are created lazily, when needed.
+What exactly means "needed" depends on the scope of a bean.
+
+* A *normal scoped bean* (`@ApplicationScoped`, `@RequestScoped`, etc.) is needed when a method is invoked upon an injected instance (contextual reference per the specifitation). 
++
+In other words, injecting a normal scoped bean is not enough because a _client proxy_ is injected instead of a contextual instance of the bean.
```
Comment (geoand) [inline docs/src/main/asciidoc/cdi-reference.adoc]: `is not enough` -> `will not suffice`

File: docs/src/main/asciidoc/cdi-reference.adoc (hunk 2 above)
Comment (gsmet) [inline docs/src/main/asciidoc/cdi-reference.adoc]:
```suggestion
* A *normal scoped bean* (`@ApplicationScoped`, `@RequestScoped`, etc.) is needed when a method is invoked upon an injected instance (contextual reference per the specification). 
```

File: docs/src/main/asciidoc/cdi-reference.adoc [hunk 4]
```diff
@@ -168,6 +168,89 @@ public class CounterBean {
...
+----
+@ApplicationScoped
+class CoolService {
+  void startup(@Observes StartupEvent event) { <1>
+  }
+}
+----
+<1> A `CoolService` is created during startup to service the observer method invocation.
+
+* Use the bean in an observer observer of the `StartupEvent` - normal scoped beans must be used as described in <<lazy_by_default>>:
```
Comment (gsmet) [inline docs/src/main/asciidoc/cdi-reference.adoc]:
```suggestion
* Use the bean in an observer of the `StartupEvent` - normal scoped beans must be used as described in <<lazy_by_default>>:    
```

Comment (geoand) [inline docs/src/main/asciidoc/cdi-reference.adoc]: `observer` is used twice

**Review by gsmet — DISMISSED** [review]
Spotted a duplicated word. Otherwise, it looks good, thanks!

File: docs/src/main/asciidoc/cdi-reference.adoc (hunk 2 above)
Comment (gsmet) [reply docs/src/main/asciidoc/cdi-reference.adoc]: This one was fixed while I was reviewing :).

**Comment by geoand** [comment]
@mkouba I wonder if it makes sense to have some kind of annotation and / or configuration what will behind the scenes result in the of the startup event to trigger eager instantiation

**Review by geoand — APPROVED** [review]
Clears things up a lot, thanks!

File: docs/src/main/asciidoc/cdi-reference.adoc (hunk 2 above)
Comment (geoand) [reply docs/src/main/asciidoc/cdi-reference.adoc]: I am learning from the master reviewer :wink:

**Comment by mkouba** [comment]
> I wonder if it makes sense to have some kind of annotation and / or configuration what will behind the scenes result in the of the startup event to trigger eager instantiation

That's a good question. Basically something like `javax.ejb.Startup`, right? I'm not sure TBH. But feel free to create a tracker issue.

**Comment by geoand** [comment]
I've never seen that annotation but yeah,  the idea is the same. I'll open an issue.

**Pull request merged**